l10n_it_fiscalcode/wizard/compute_fc.py

```diff
@@ -1,54 +1,110 @@
-import datetime
 
 _logger = logging.getLogger(__name__)
 
 try:
     from codicefiscale import build
+
 except ImportError:
     _logger.warning(
-        'codicefiscale library not found. '
-        'If you plan to use it, please install the codicefiscale library '
-        'from https://pypi.python.org/pypi/codicefiscale')
+        "codicefiscale library not found. "
+        "If you plan to use it, please install the codicefiscale library"
+        " from https://pypi.python.org/pypi/codicefiscale")
 
 
 class WizardComputeFc(models.TransientModel):
-
-    _name = "wizard.compute.fc"
+    _name = 'wizard.compute.fc'
     _description = "Compute Fiscal Code"
     _rec_name = 'fiscalcode_surname'
 
-    fiscalcode_surname = fields.Char('Surname', size=64)
-    fiscalcode_firstname = fields.Char('First name', size=64)
-    birth_date = fields.Date('Date of birth')
-    birth_city = fields.Many2one(
-        'res.city.it.code.distinct', string='City of birth')
-    birth_province = fields.Many2one(
-        'res.city.it.code.province', string='Province')
-    sex = fields.Selection([
-        ('M', 'Male'),
-        ('F', 'Female'),
-        ], "Sex")
+    fiscalcode_surname = fields.Char("Surname", required=True, size=64)
+    fiscalcode_firstname = fields.Char("First name", required=True, size=64)
+    birth_date = fields.Date("Date of birth", required=True)
+    birth_city = fields.Many2one('res.city.it.code.distinct',
+                                 required=True,
+                                 string="City of birth")
+    birth_province = fields.Many2one('res.country.state',
+                                     required=True,
+                                     string="Province")
+    sex = fields.Selection([('M', "Male"), ('F', "Female")],
+                           required=True,
+                           string="Sex")
 
     @api.multi
     @api.onchange('birth_city')
     def onchange_birth_city(self):
         self.ensure_one()
-        res = {}
+
+        it = self.env.ref('base.it').id
+        res = {
+            'domain': {'birth_province': [('country_id', '=', it)]},
+            'value': {'birth_province': False}
+        }
+
         if self.birth_city:
-            ct = self.birth_city
-            res['domain'] = {
-                'birth_province': [('town_name', '=', ct.name)]
-            }
-        else:
-            res['domain'] = {'birth_province': []}
-        res['value'] = {'birth_province': ''}
+            city = self.birth_city.name
+
+            # SMELLS: Add a foreign key in "res_city_it_code"
+            #          instead using the weak link "code" <-> "province".
+            #
+            self.env.cr.execute("""
+                SELECT "p"."id" FROM "res_country_state" AS "p"
+                    INNER JOIN "res_city_it_code" AS "c"
+                        ON ("p"."code" = "c"."province")
+
+                WHERE "p"."country_id" = %s AND "c"."name" = %s;
+            """, (it, city))
```

Comment:
@Byloth Come mai SQL invece dell'ORM?

Reply:
Questa semplice SQL, sarebbe dovuta diventare qualcosa del tipo:
```python
city_ids = self.env['res.city.it.code'].search([('name', '=', city)])
codes = city_ids.mapped('code')
province_ids = self.env['res.country.state'].search([('country_id', '=', it), ('province', 'in', codes)])
```
---

Personalmente, trovo molto più intuitiva la SQL anziché il giro fatto con l'ORM (escludendo, indubbiamente, poi il discorso legato alle prestazioni...).

Se, però, ritieni sia meglio l'uso dell'ORM, fixo subito! 😊

Reply:
Usando SQL invece dell'ORM, eventuali regole di sicurezza, sia a livello di tabella che di riga, vengono ignorate.
In questo caso non dovrebbe capitare, ma considera che tali regole potrerbbero anche essere aggiunte manualmente dagli utenti amministratori.

Relativamente all'intuitività, io mi trovo meglio con python piuttosto che SQL, ma sono gusti :-)
Sono comunque 3 righe contro 8

Reply:
Indipendentemente dalle permissions configurate dei singoli amministratori, in questi due casi specifici, non avrei comunque interferito con tali regole in quanto vado ad utilizzare i risultati delle query solo per impostare il valore degli attributi `domain` sui campi che poi, successivamente, vengono utilizzati.  
Solo a quel punto sarebbero entrate in gioco le permissions più disparate (così com'è giusto che sia). 😉

In ogni caso, ho appena corretto! 😁